### PR TITLE
Add cluster intelligence collection subcommand

### PR DESCRIPTION
**Pull request by StefanHauth**
# Description

Provides a subcommand to collect useful information for troubleshooting:
- logs of operator related pods
- manifests deployed by the operator and related manifests
- operator version

The information is packed into a zipped tarball.

## How can this be tested?
Deploy the operator and run
```
kubectl -n dynatrace exec pod/$(kubectl -n dynatrace get pods | grep dynatrace-operator | cut -f1 -d' ') -- /usr/local/bin/dynatrace-operator cic
```
As the last line of output you will get a kubectl command that can be used to copy the resulting tarball from the pod to your local machine.

## Checklist
- [x] Unit tests have been updated/added
- [x] PR is labeled accordingly
- [x] I have read and understood the [contribution guidelines](https://github.com/Dynatrace/dynatrace-operator/blob/master/CONTRIBUTING.md)

### CODE REVIEW

File: config/helm/chart/default/templates/Common/operator/clusterrole-operator.yaml [hunk 1]
```diff
@@ -28,6 +28,12 @@ rules:
       - get
       - list
       - watch
+  - apiGroups:
+      - ""
+    resources:
+      - pods/log
+    verbs:
+      - get
```
Comment (chrismuellner) [inline config/helm/chart/default/templates/Common/operator/clusterrole-operator.yaml:36]: since we are only accessing operator components with known names, can this be restricted to specific `resourceNames`?

File: src/cmd/cluster_intel_collector/manifest_collector.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
+package cluster_intel_collector
```
Comment (meik99) [inline src/cmd/cluster_intel_collector/manifest_collector.go:1]: the name of this file could just be `manifests.go`

File: src/cmd/cluster_intel_collector/tarball.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
+package cluster_intel_collector
```
Comment (meik99) [inline src/cmd/cluster_intel_collector/tarball.go:1]: there are no unit tests for this file.
if there are specific unit tests they should be in `tarball_test.go`

File: src/cmd/cluster_intel_collector/version_collector.go [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+	"fmt"
+	"runtime"
+	"strings"
+
+	"github.com/Dynatrace/dynatrace-operator/src/version"
+)
+
+const versionFileName = "operator-version.txt"
+
+func collectOperatorVersion(_ *intelCollectorContext, tarball *intelTarball) error {
```
Comment (meik99) [inline src/cmd/cluster_intel_collector/version_collector.go:13]: why have an `intelCollectorContext` parameter if it is not used?

File: src/cmd/cluster_intel_collector/version_collector.go [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
```
Comment (meik99) [inline src/cmd/cluster_intel_collector/version_collector.go:14]:
```suggestion
```

File: src/cmd/cluster_intel_collector/log_collector.go [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+func collectLogs(cicCtx *intelCollectorContext, tarball *intelTarball) error {
+	podList, err := getPodList(cicCtx)
+	if err != nil {
+		return err
+	}
+
+	for _, pod := range podList.Items {
+		podGetOptions := metav1.GetOptions{}
+		pod, err := cicCtx.clientSet.CoreV1().Pods(cicCtx.namespaceName).Get(cicCtx.ctx, pod.Name, podGetOptions)
```
[the remark this answers is not in the document]
Comment (0sewa0) [reply src/cmd/cluster_intel_collector/log_collector.go:20]: I don't think this is the client that we use, this is the what is inside our `client.Client` that is provided by the controller-runtime (I think)

File: src/cmd/cluster_intel_collector/version_collector.go (hunk 5 above)
Comment (meik99) [reply src/cmd/cluster_intel_collector/version_collector.go:14]: just to avoid confusion here, there is an empty line that can be removed

File: src/cmd/cluster_intel_collector/version_collector.go (hunk 4 above)
Comment (StefanHauth) [reply src/cmd/cluster_intel_collector/version_collector.go:13]: Because there's a common function definition for all collector functions which is used in main.

File: src/cmd/cluster_intel_collector/builder.go [hunk 7]
```diff
@@ -0,0 +1,126 @@
+package cluster_intel_collector
```
[the remark this answers is not in the document]
Comment (StefanHauth) [reply src/cmd/cluster_intel_collector/builder.go:1]: Why avoid speaking package names just because they are a bit longer? Although I'm not completely convinced by `cluster_intel_collector` myself. just `support` or `archive` or `collector` are too generic for my taste. I'd vote for `support_archive`, also because it is (as mentioned in a comment below) an well established concept.

File: src/cmd/cluster_intel_collector/builder.go [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+	"github.com/Dynatrace/dynatrace-operator/src/cmd/config"
+	"github.com/Dynatrace/dynatrace-operator/src/scheme"
+	"github.com/spf13/cobra"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/cluster"
+)
+
+const (
+	use               = "cic"
```
[the remark this answers is not in the document]
Comment (StefanHauth) [reply src/cmd/cluster_intel_collector/builder.go:18]: Good idea, will do that!

File: src/cmd/cluster_intel_collector/builder.go [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+	return CommandBuilder{}
+}
+
+func (builder CommandBuilder) SetConfigProvider(provider config.Provider) CommandBuilder {
+	builder.configProvider = provider
+	return builder
+}
+
+func (builder CommandBuilder) GetCluster(kubeConfig *rest.Config) (cluster.Cluster, error) {
+	return cluster.New(kubeConfig, clusterOptions)
```
[the remark this answers is not in the document]
Comment (StefanHauth) [reply src/cmd/cluster_intel_collector/builder.go:44]: Should this apply to troubleshoot as well? That's where I "got my inspiration" :-)

File: src/cmd/cluster_intel_collector/log_collector.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (StefanHauth) [reply src/cmd/cluster_intel_collector/log_collector.go:20]: The log collector has to use the client set anyway to get logs. So that's what used to do all API calls in this file to keep it consistent.

File: src/cmd/cluster_intel_collector/log_collector.go [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+
+	switch {
+	case logOptions.Previous && err != nil:
+		// Soften error message for previous pods a bit, so users don't get nervous. Previous pods often just didn't exist.
+		logInfof("logs for previous pod not found: %v", err)
+		return
+	case err != nil:
+		logErrorf("error in opening stream: %v", err)
+		return
+	}
```
[the remark this answers is not in the document]
Comment (StefanHauth) [reply src/cmd/cluster_intel_collector/log_collector.go:69]: IMHO switches provide better readability.

File: src/cmd/cluster_intel_collector/log_collector_test.go [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+	tarball := intelTarball{
+		tarWriter: tar.NewWriter(&tarBuffer),
+	}
+
+	require.NoError(t, collectLogs(&ctx, &tarball))
+	tarball.tarWriter.Close()
+
+	tarReader := tar.NewReader(&tarBuffer)
+
+	hdr, err := tarReader.Next()
```
[the remark this answers is not in the document]
Comment (StefanHauth) [reply src/cmd/cluster_intel_collector/log_collector_test.go:104]: `header`, but I'll make that clearer

File: src/cmd/cluster_intel_collector/manifest_collector.go [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+
+import (
+	"bytes"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+func collectManifests(cicCtx *intelCollectorContext, tarBall *intelTarball) error {
```
[the remark this answers is not in the document]
Comment (StefanHauth) [reply src/cmd/cluster_intel_collector/manifest_collector.go:10]: There are several different collectors that collect different things. The context is just a class that holds some stuff that's needed by all of them, but the context itself is not doing any collection work. So, no, neither these nor other functions related to collecting stuff should be member functions of the context.

File: src/cmd/cluster_intel_collector/config.go [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+
+var (
+	log = newLogCollectorLogger("[log collector]")
+)
+
+type intelCollectorContext struct {
+	ctx           context.Context
+	clientSet     kubernetes.Interface // used to get access to logs
+	apiReader     client.Reader        // used for manifest collection
+	namespaceName string               // the default namespace ("dynatrace") or provided in the command line
```
[the remark this answers is not in the document]
Comment (StefanHauth) [reply src/cmd/cluster_intel_collector/config.go:19]: The comments (should) explain to the less experienced why two kinds of clients have to be used instead of just one.

File: src/cmd/cluster_intel_collector/config.go [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/kubernetes"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var (
+	log = newLogCollectorLogger("[log collector]")
+)
+
+type intelCollectorContext struct {
```
[the remark this answers is not in the document]
Comment (StefanHauth) [reply src/cmd/cluster_intel_collector/config.go:15]: We have different kind of collectors, packing them all into one type isn't good software design. The context provides access to general stuff all of the collectors need. Otherwise the collectors shall be kept separated. The context itself fullfills no collection tasks whatever.

File: src/cmd/cluster_intel_collector/config.go [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+	ctx           context.Context
+	clientSet     kubernetes.Interface // used to get access to logs
+	apiReader     client.Reader        // used for manifest collection
+	namespaceName string               // the default namespace ("dynatrace") or provided in the command line
+	toStdout      bool
+	targetDir     string
+}
+
+type manifestSpec struct {
```
[the remark this answers is not in the document]
Comment (StefanHauth) [reply src/cmd/cluster_intel_collector/config.go:24]: Agreed, was not happy with `manifestSpec` either. But I'll move the list of object queries into a separate file, cause it just makes the `manifests.go` file unnecessarily long otherwise.

File: src/cmd/cluster_intel_collector/log_collector.go [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+import (
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const tarFileName = "%s/operator-cic-%s.tgz"
+
+func collectLogs(cicCtx *intelCollectorContext, tarball *intelTarball) error {
```
[the remark this answers is not in the document]
Comment (StefanHauth) [reply src/cmd/cluster_intel_collector/log_collector.go:12]: As before, the context is not doing collection work, it's a supporting type for collection functions of all kinds. Therefore it belongs into it's own file and it doesn't have member functions for collecting stuff.

File: src/cmd/cluster_intel_collector/tarball.go [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+	"compress/gzip"
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+	"strings"
+	"time"
+)
+
+type intelTarball struct {
```
[the remark this answers is not in the document]
Comment (StefanHauth) [reply src/cmd/cluster_intel_collector/tarball.go:15]: The term `tarWriter` is already used by the underlying lib, makes it confusing. And the tarball is more than just the tarWriter.

File: src/cmd/cluster_intel_collector/logger.go [hunk 18]
```diff
@@ -0,0 +1,93 @@
...
+	config.NameKey = "name"
+	config.EncodeTime = zapcore.ISO8601TimeEncoder
+
+	return logr.New(
+		logCollectorLogger{
+			// using stderr because we use stdout to deliver the tarball
+			logger: ctrlzap.New(ctrlzap.WriteTo(os.Stderr), ctrlzap.Encoder(zapcore.NewConsoleEncoder(config))).WithName(testName),
+		},
+	)
+}
```
[the remark this answers is not in the document]
Comment (StefanHauth) [reply src/cmd/cluster_intel_collector/logger.go:40]: The problem here is, that DTLogger has (yet) no possibility to use a different encoding than JSON. While JSON is fine in "normal" operation, it's not a good choice for the usability of a CLI command.

File: src/cmd/cluster_intel_collector/builder.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (0sewa0) [reply src/cmd/cluster_intel_collector/builder.go]: so the problem is that the whole structure is bit off.

having a separate command for the current `troubleshoot` and the `collection` is what we want, they should probably be sub commands of a `support` or `troubleshoot` command

so how it would work in my head is that we have a `support` command (which would also the name of the package)
then there would be a sub package called `troubleshoot` where the current `troubleshoot` package should be moved
and the contents of this PR would be moved into an `archive` package
so the end result would look like:
```
/src/cmd/support
                    |
                    |- troubleshoot/
                    |                 |
                    |           troubleshoot.go
                    |            ....
                    |
                    |- archive/
                    |         |
                    |        archive.go
                    |         ....
```
and from the user’s perspective there would be a `support` command with sub commands like `archive` and `troubleshoot`
examples: `support archive` or `support troubleshoot`

this would allow us to combine/chain the commands, so the user/support person has to run a single command that under the hood uses multiple sub commands.
example: `support` command without subcommands would
- Run the `troubleshoot` subcommand, and store the result somewhere
- Run the log `archive` subcommand, which would also put in the archive the result of the  `troubleshoot` subcommand


This how I think of it at least. This should be discussed with the team, and make sure that we agree on a design that works for all

TLDR: The current solution feels very rushed, and I think we need to go back to the drawing board, if we want  to introduce a feature that we wish to support and improve on in the future.

[the remark this answers is not in the document]
Comment (StefanHauth) [reply src/cmd/cluster_intel_collector/builder.go]: Ok, we can discuss this, there are some pitfalls if we want to combine those commands into one (which it essentially is, if we put it under one common subcommand). But that is definitely out of scope for this PR.

File: src/cmd/cluster_intel_collector/tarball.go (hunk 3 above)
Comment (StefanHauth) [reply src/cmd/cluster_intel_collector/tarball.go]: The tarball is indirectly part of the log, objects and version collection tests. I'm aware that this way tests are not strictly limited to one unit, OTH unit tests for the tarball only would very much look the same as those tests, so it feels redundant.

**Comment by StefanHauth** [comment]
Closing PR, back to drawing board.
